File: [J]backtracking/[J]backtracking/79-word-search.java
class Solution {
    char[][] board;
    String word;
    int rows;
    int cols;

    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.word = word;
        rows = board.length;
        cols = board[0].length;

        HashMap<Character, Integer> charToFreq = new HashMap<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                charToFreq.compute(board[r][c], (k, v) -> v == null ? 1 : v + 1);
            }
        }
        for (int i = 0; i < word.length(); i++) {
            Character c = word.charAt(i);
            if (!charToFreq.containsKey(c)) {
                return false;
            }
            charToFreq.put(c, charToFreq.get(c) - 1);
            if (Objects.equals(charToFreq.get(c), 0)) {
                charToFreq.remove(c);
            }
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                HashSet<Integer> visited = new HashSet<>();
                if (dfs(r, c, 0, visited) == true) {
                    return true;
                }
            }
        }
        return false;
    }

    public boolean dfs(int r, int c, int idx, HashSet<Integer> visited) {
        if (0 > r || r >= rows || 0 > c || c >= cols || visited.contains(r * cols + c) || board[r][c] != word.charAt(idx)) {
            return false;
        }
        if (idx == word.length() - 1) {
            return true;
        }
        visited.add(r * cols + c);
        for (int[] nextrc: new int[][]{{r+1, c}, {r-1, c}, {r, c+1}, {r, c-1}}) {
            if (dfs(nextrc[0], nextrc[1], idx + 1, visited) == true) {
                return true;
            }
        }
        visited.remove(r * cols + c);
        return false;

    }
}

// time O(RC * 3**L)
// space O(RC), due to hashset
// using dfs and backtracking and backtracking with constraints and pruning

File: [J]backtracking/[J]backtracking/79-word-search.py
from collections import defaultdict
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])

        schar_freq = defaultdict(int)
        for r in range(rows):
            for c in range(cols):
                schar_freq[board[r][c]] += 1
        tchar_freq = defaultdict(int)
        for c in word:
            tchar_freq[c] += 1
        for tc, freq in tchar_freq.items():
            if freq > schar_freq[tc]:
                return False
        
        def dfs(r, c, idx, visited):
            if idx == len(word):
                return True
            for next_r, next_c in [(r+1, c), (r-1, c), (r, c+1), (r, c-1)]:
                if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited or board[next_r][next_c] != word[idx]:
                    continue
                visited.add((next_r, next_c))
                if dfs(next_r, next_c, idx + 1, visited):
                    return True
                visited.remove((next_r, next_c))
            return False

        for r in range(rows):
            for c in range(cols):
                if board[r][c] == word[0]:
                    if dfs(r, c, 1, {(r, c)}):
                        return True
        return False
    
# time O(RC * 3**L)
# space O(RC), due to hashset
# using dfs and backtracking and backtracking with constraints and pruning